### PR TITLE
Improvements to instrument names translations

### PR DESCRIPTION
**Pull request by cbjeukendrup**
- Added disambiguations when translating instrument names, per instrument id and longName/shortName/trackName
- Merged generateTs.py into update_instruments_xml.py

### CODE REVIEW

File: share/instruments/instrumentsxml.h [hunk 1]
```diff
@@ -189,2502 +195,2918 @@ QT_TRANSLATE_NOOP("InstrumentsXML", "Viols"),
...
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Picc.", "db-piccolo|shortName"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Piccolo", "db-piccolo|trackName"),
+//: traitName
+QT_TRANSLATE_NOOP("InstrumentsXML", "D♭"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Piccolo in D♭, sounding a semitone above the standard piccolo.", "db-piccolo|description"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Piccolo", "piccolo|longName"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Picc.", "piccolo|shortName"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Piccolo", "piccolo|trackName"),
+//: traitName
+QT_TRANSLATE_NOOP("InstrumentsXML", "*(C)"),
```
Comment (shoogle) [inline share/instruments/instrumentsxml.h:223]: It was clever to use a translation comment like this but I think we will need disambiguation text even for the trait names. The reason is that we add special characters to the trait names to give them special meanings:

`*` - This is the most common trait name so use it as the default in the New Score Wizard dropdown.
`()` - This trait name should not appear in the score.

These are added to the translatable string to allow translators to choose a different default to use in their locale, and to hide or unhide the trait name in scores written in their language.

If the `//:` comment strings are actually exposed to translators then you could make use of it like this:

```c++
//: traitName codes: * most common for instrument; () hide in score
QT_TRANSLATE_NOOP3("InstrumentsXML", "*(C)", "piccolo|traitName"),
```

A translator could choose to translate `"*(C)"` for the piccolo as `"C"`, `"*C"`, `"(C)"` or `"*(C)"`, and they might make a different choice for a different instrument that also has `"*(C)"` as the trait name in English, hence we need disambiguation text.

**Review by shoogle — CHANGES_REQUESTED** [review]

File: share/instruments/instrumentsxml.h [hunk 2]
```diff
@@ -189,2502 +195,2918 @@ QT_TRANSLATE_NOOP("InstrumentsXML", "Viols"),
...
-QT_TRANSLATE_NOOP("InstrumentsXML", "Bass"),
-QT_TRANSLATE_NOOP("InstrumentsXML", "Kazoo"),
-QT_TRANSLATE_NOOP("InstrumentsXML", "Kazoo"),
-QT_TRANSLATE_NOOP3("InstrumentsXML", "Kaz.", "Kazoo"),
-QT_TRANSLATE_NOOP("InstrumentsXML", "Kazoo"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Boy Soprano", "boy-soprano|longName"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "B. S.", "boy-soprano|shortName"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Boy Soprano", "boy-soprano|trackName"),
+QT_TRANSLATE_NOOP3("InstrumentsXML", "Boy soprano voice.", "boy-soprano|description"),
+QT_TRANSLATE_NOOP3("
```
[the remark this answers is not in the document]
Comment (cbjeukendrup) [reply share/instruments/instrumentsxml.h]: Oops, that was by accident, by my git pre-commit hook. It just uncrustified every staged file. Fixed!

File: src/engraving/libmscore/instrtemplate.cpp [hunk 3]
```diff
@@ -432,6 +432,12 @@ void InstrumentTemplate::write1(XmlWriter& xml) const
 //   read
 //---------------------------------------------------------
 
+static QString translateInstrumentName(const QString& instrumentId, const QString& nameType, const QString& text)
+{
+    QString disambiguation = instrumentId + '|' + nameType;
+    return qtrc("InstrumentsXML", text.toUtf8().data(), disambiguation.toUtf8().data());
+}
```
[the remark this answers is not in the document]
Comment (cbjeukendrup) [reply src/engraving/libmscore/instrtemplate.cpp:439]: I'm not sure what to do about channels. It seems that before, they were not translated at all. I also believe that the whole idea of channels is being dropped anyway.

[the remark this answers is not in the document]
Comment (shoogle) [reply src/engraving/libmscore/instrtemplate.cpp:439]: Let's keep them in the translation file for now so we don't have to add them again later after string freeze. There is only a relatively small number of them anyway.

BTW, my understanding is that channels have been dropped for MuseScore 4.0 to save time, but they may return in 4.1 in some form, though they may work very differently to before.

File: share/instruments/update_instruments_xml.py [hunk 4]
```diff
@@ -319,13 +327,116 @@ def to_list(str):
...
+            instrumentId = instrument['id']
+
+            if instrument['traitName'] != '[hide]':
+                for nameType in ['longName', 'shortName', 'trackName']:
+                    add_translatable_string_if_not_null(f, 'InstrumentsXML', instrument[nameType],
+                                                        disambiguation(instrumentId, nameType))
+                
+                add_translatable_string_if_not_null(f, 'InstrumentsXML', instrument['traitName'],
+                                                    disambiguation(instrumentId, 'traitName'), 
+                                                    'traitName codes: * most common for instrument; () hide in score')
```
Comment (shoogle) [inline share/instruments/update_instruments_xml.py:429]: Could you write 'longName', 'shortName' and 'trackName' even if the trait name is '[hide]'? We may need translations of those strings to identify instruments in older scores that don't have instrument IDs. I think we only started saving instrument IDs in MuseScore 3.5 or thereabouts, so in older scores we look at the instrument name. We're not doing this yet but there was a plan to do so, so we'll need them in before the string freeze.

Obviously we don't to translate the trait name if it is '[hide]', so you can keep that in the conditional statement.

**Review by shoogle — CHANGES_REQUESTED** [review]

**Review by shoogle — APPROVED** [review]
Awesome!

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**